src/v/archival/segment_reupload.cc

```diff
@@ -327,6 +327,19 @@ segment_collector::make_upload_candidate(
         co_return upload_candidate_with_locks{upload_candidate{}, {}};
     }
 
+    // Take the locks before opening any readers on the segments.
+    auto deadline = std::chrono::steady_clock::now() + segment_lock_duration;
```

Comment:
Not related to this PR, but there is a bug (or at least a misnomer) carried forward here from the original version of the code, the segment lock duration is not actually a thing governing lock's lifetime. It is the time spent waiting to take the lock, if the semaphore units are not obtained by the deadline the lock is not granted and an exception is thrown. 

Once the lock is taken however, it can be held indefinitely, it does not expire after `segment_lock_duration` but only when the object is destroyed. 

I do not think there is any way to have expiring locks other than adding a separate system to keep track of them.

Reply:
Thanks for this detail! It is indeed surprising (at least I didn't expect that)